src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,513 @@
+{
+    auto const present = steady_clock::now();
+
+    Json::Value jobsArray(Json::arrayValue);
+    decltype(jobs_) jobs;
+    {
+        std::lock_guard<std::mutex> lock(jobsMutex_);
+        jobs = jobs_;
+    }
+
+    for (auto const& j : jobs)
+    {
+        if (j.first == jtINVALID)
+            continue;
+        Json::Value jobj(Json::objectValue);
+        auto const e = jq_.find(j.first);
+        if (e == jq_.end())
+        {
+            assert(false);
+            continue;
+        }
+        {
+            std::lock_guard<std::mutex> lock(e->second.mut);
+            jobj[jss::job] = e->second.label;
+        }
+        jobj[jss::duration_us] = std::to_string(
+            std::chrono::duration_cast<microseconds>(
+                present - j.second).count());
+        jobsArray.append(jobj);
+    }
+
+    Json::Value methodsArray(Json::arrayValue);
+    decltype(methods_) methods;
+    {
+        std::lock_guard<std::mutex> lock(methodsMutex_);
+        methods = methods_;
+    }
+    for (auto m : methods)
+    {
+        Json::Value methodobj(Json::objectValue);
+        methodobj[jss::method] = m.second.first;
+        methodobj[jss::duration_us] = std::to_string(
+            std::chrono::duration_cast<microseconds>(
+                present - m.second.second).count());
+        methodsArray.append(methodobj);
+    }
+
+    Json::Value current(Json::objectValue);
+    current[jss::jobs] = jobsArray;
+    current[jss::methods] = methodsArray;
+    return current;
+}
+
+//-----------------------------------------------------------------------------
+
+void
+PerfLogImp::openLog()
+{
+    if (! setup_.perf_log.empty())
+    {
+        if (logFile_.is_open())
+            logFile_.close();
+
+        auto log_dir = setup_.perf_log.parent_path();
+        if (!boost::filesystem::is_directory(log_dir))
+        {
+            boost::system::error_code ec;
+            boost::filesystem::create_directories(log_dir, ec);
+            if (ec)
+            {
+                journal_.fatal() << "Unable to create performance log "
+                    "directory " << log_dir << ": " << ec.message();
+                app_.signalStop();
+                return;
+            }
+        }
+
+        try
+        {
+            logFile_.open(setup_.perf_log.c_str(),
+                std::ios::out | std::ios::app);
+        }
+        catch (std::exception const& e)
+        {
+            journal_.fatal() << "Unable to open performance log " <<
+                setup_.perf_log << ": " << e.what();
+            app_.signalStop();
+        }
+    }
+}
+
+void
+PerfLogImp::run()
+{
+    beast::setCurrentThreadName("perflog");
+    lastLog_ = system_clock::now();
+
+    while (true)
+    {
+        {
+            std::unique_lock<std::mutex> lock(mutex_);
+            if (stop_)
+            {
+                stopped();
+                return;
+            }
+            if (rotate_)
+            {
+                openLog();
```

Comment:
Log rotation involves renaming the file at the OS level, but the filehandle needs to be closed and then reopened for a new file to be created in the original location. Otherwise the original file with the new name is continually appended to. What you're not seeing is the file actually getting rotated, which is performed externally to rippled.

Reply:
So the process is:

1. From some process other than rippled rename the log file.  This can't take effect yet, since rippled has the file open.
2. Tell rippled to `log_rotate`.  When the rotate happens, rippled closes the file which allows the rename (from step 1) to complete.
3. Then rippled opens the log file it was writing to before.  Since the rename took effect after the previous close, rippled creates a new file with the same (old) log file name.  All future logging goes into this newly opened file.

Is that kind of what's happening?  Thanks.

Reply:
@scottschurr 
In UNIX file systems, the name of the file is not the same as the file handle. The pathname is used by the application to find the file for opening. Renaming the file is simply changing a record in the file's header record known as an inode. If the directory changes, then the source and destination directories' inodes are modified accordingly. The application knows no difference at this point: it keeps writing/reading/seeking the same exact file, but which has a different pathname. Closing the filehandle stops these activities. Opening again with the original pathname creates a new file in the original location.

Reply:
Thanks for the great explanation.  It will help me internalize why renaming a file in UNIX is called `mv` :smile:.  Long term we may need to think about whether what we're doing works on Windows.  And @mDuo13 may want to document how we expect folks to use `log_rotate`, as well as what it does.  But I'm good.  Thanks for the help.